File: test_spotlight.py
from playwright.sync_api import sync_playwright
import time
import os

def test_spotlight_guide():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context()
        page = context.new_page()
        
        screenshot_dir = r"C:\Users\wyxgg\Desktop\test\test_screenshots"
        os.makedirs(screenshot_dir, exist_ok=True)
        
        def screenshot(name):
            time.sleep(0.3)
            path = os.path.join(screenshot_dir, f"spotlight_{name}.png")
            page.screenshot(path=path)
            print(f"[screenshot] spotlight_{name}.png")
        
        try:
            print("=== Test Spotlight Guide ===\n")
            
            # 1. 打开首页
            page.goto("http://localhost:3000")
            page.wait_for_timeout(500)
            screenshot("01_homepage")
            print("[OK] 首页加载完成")
            
            # 2. 清除 localStorage
            page.evaluate("localStorage.clear()")
            page.reload()
            page.wait_for_timeout(500)
            screenshot("02_after_clear_storage")
            
            # 3. 登录框应该已经打开，检查并切换到注册标签
            modal_open = page.evaluate("""
                () => {
                    const modal = document.getElementById('loginModal');
                    return modal ? getComputedStyle(modal).display !== 'none' : false;
                }
            """)
            print(f"登录框已打开: {modal_open}")
            page.wait_for_timeout(300)
            screenshot("03_login_modal")
            
            # 4. 切换到注册标签
            page.click("button.login-tab:text('注册')")
            page.wait_for_timeout(300)
            screenshot("04_register_tab")
            
            # 5. 填写注册信息
            import random
            username = f"testuser_{int(time.time())}"
            page.fill("#regUsernameInput", username)
            page.fill("#regPasswordInput", "test123456")
            # 昵称需要2-12字
            page.fill("#regNicknameInput", f"测试用户{int(time.time()) % 10000}")
            screenshot("05_register_filled")
            print(f"注册用户: {username}")
            
            # 6. 点击登录/注册按钮（在注册Tab下会调用doUserRegister）
            page.click("#loginConfirmBtn")
            page.wait_for_timeout(1500)
            screenshot("06_after_register")
            
            # 7. 检查引导是否出现
            page.wait_for_timeout(1000)  # 等待引导触发
            
            # 先检查 _spotlight 对象状态
            spotlight_debug = page.evaluate("""
                () => {
                    return {
                        spotlightExists: typeof _spotlight !== 'undefined',
                        spotlightStep: typeof _spotlight !== 'undefined' ? _spotlight.step : null,
                        spotlightStepsLen: typeof _spotlight !== 'undefined' ? _spotlight.steps.length : null,
                        currentUser: typeof currentUser !== 'undefined' ? currentUser : null,
                        guideShown: localStorage.getItem('newbieGuideShown'),
                        userKey: localStorage.getItem('campus_wall_user') ? 'exists' : 'null',
                        forceOff: localStorage.getItem('newbieGuideForceOff')
                    };
                }
            """)
            print(f"Debug: {spotlight_debug}")
            
            guide_visible = page.evaluate("""
                () => {
                    const overlay = document.getElementById('spotlightOverlay');
                    const tooltip = document.getElementById('spotlightTooltip');
                    if (!overlay || !tooltip) return { found: false };
                    return {
                        found: true,
                        overlayDisplay: getComputedStyle(overlay).display,
                        tooltipDisplay: getComputedStyle(tooltip).display,
                        tooltipText: tooltip ? tooltip.innerText : ''
                    };
                }
            """)
            
            if not guide_visible.get('found'):
                print("[FAIL] 引导元素未找到！")
                return
            
            if guide_visible['tooltipDisplay'] == 'none':
                print("[FAIL] 引导未显示！")
                screenshot("07_guide_not_shown")
                return
            
            print("[OK] 引导已出现！")
            print(f"引导内容: {guide_visible['tooltipText'].replace(chr(10), ' | ')}")
            screenshot("07_guide_step1")
            
            # 获取高亮元素
            highlighted = page.evaluate("""
                () => {
                    const overlay = document.getElementById('spotlightOverlay');
                    if (!overlay) return null;
                    const svg = overlay.querySelector('svg');
                    if (!svg) return null;
                    const ring = svg.querySelector('.spotlight-ring');
                    if (!ring) return null;
                    const rect = ring.getBoundingClientRect();
                    const centerX = rect.x + rect.width / 2;
                    const centerY = rect.y + rect.height / 2;
                    const el = document.elementFromPoint(centerX, centerY);
                    return el ? (el.className + ' #' + el.id) : null;
                }
            """)
            print(f"高亮元素: {highlighted}")
            
            # 8. 检查按钮文字
            btn_text = page.locator("button.spotlight-next, button.spotlight-done").first.text_content()
            print(f"引导按钮: {btn_text}")
            print(f"引导按钮: {btn_text}")
            
            # 9. 点"知道了"打开发帖模态框
            page.click("button.spotlight-next, button.spotlight-done")
            page.wait_for_timeout(800)
            screenshot("08_post_modal_opened")
            
            # 检查模态框是否打开
            modal_open = page.evaluate("""
                () => getComputedStyle(document.getElementById('modalOverlay')).display !== 'none'
            """)
            modal_visible = page.evaluate("getComputedStyle(document.getElementById('modalOverlay')).display !== 'none'")
            print(f"发帖模态框已打开: {modal_visible}")
            
            # 10. 填写帖子内容
            page.fill("#noteInput", "这是引导测试帖子！")
            screenshot("09_post_filled")
            
            # 11. 提交帖子
            page.click("button.modal-submit")
            page.wait_for_timeout(1500)
            screenshot("10_after_post")
            
            # 12. 检查引导状态
            guide_after = page.evaluate("""
                () => {
                    const tooltip = document.getElementById('spotlightTooltip');
                    return tooltip ? {
                        visible: getComputedStyle(tooltip).display !== 'none',
                        text: tooltip.innerText.replace(/\\n/g, ' | ')
                    } : null;
                }
            """)
            
            if guide_after and guide_after['visible']:
                print(f"发帖后引导: {guide_after['text']}")
                screenshot("11_guide_step2")
            else:
                print("[WARN] 发帖后引导已关闭")
                screenshot("11_guide_closed")
            
            print("\n[OK] 测试完成！")
            
        except Exception as e:
            print(f"[FAIL] 测试失败: {e}")
            screenshot("error")
        finally:
            browser.close()

if __name__ == "__main__":
    test_spotlight_guide()

File: _archive/test_spotlight.js
const { chromium } = require('playwright');

(async () => {
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext();
  const page = await context.newPage();

  // 截图辅助函数
  const screenshot = async (name) => {
    await page.waitForTimeout(300);
    await page.screenshot({ path: `test_spotlight_${name}.png` });
    console.log(`📸 截图: test_spotlight_${name}.png`);
  };

  try {
    console.log('=== 测试聚光灯引导流程 ===\n');

    // 1. 打开首页
    await page.goto('http://localhost:3000');
    await page.waitForTimeout(500);
    await screenshot('01_homepage');
    console.log('✅ 首页加载完成');

    // 2. 清除 localStorage 确保引导可以触发
    await page.evaluate(() => localStorage.clear());

    // 3. 点击登录按钮
    await page.click('#loginBtn');
    await page.waitForTimeout(300);
    await screenshot('02_login_modal');

    // 4. 切换到注册标签
    await page.click('button.login-tab:has-text("注册")');
    await page.waitForTimeout(300);
    await screenshot('03_register_tab');

    // 5. 填写注册信息
    const timestamp = Date.now();
    const username = `testuser_${timestamp}`;
    await page.fill('input[id="regUsername"]', username);
    await page.fill('input[id="regPassword"]', 'test123456');
    await page.fill('input[id="regPassword2"]', 'test123456');
    await screenshot('04_register_filled');

    // 6. 点击注册按钮
    await page.click('button:has-text("注册")');
    await page.waitForTimeout(1000);
    await screenshot('05_after_register');

    // 7. 检查引导是否出现（应该是step 1，发帖按钮）
    const spotlightVisible = await page.evaluate(() => {
      const overlay = document.getElementById('spotlightOverlay');
      const tooltip = document.getElementById('spotlightTooltip');
      return {
        overlayDisplay: overlay ? overlay.style.display : 'null',
        tooltipDisplay: tooltip ? tooltip.style.display : 'null',
        overlayVisible: overlay ? getComputedStyle(overlay).display !== 'none' : false,
        tooltipVisible: tooltip ? getComputedStyle(tooltip).display !== 'none' : false
      };
    });
    console.log('聚光灯状态:', spotlightVisible);

    if (!spotlightVisible.overlayVisible) {
      console.log('❌ 引导未出现！检查 Console 错误');
      // 检查控制台错误
      const consoleErrors = [];
      page.on('console', msg => {
        if (msg.type() === 'error') consoleErrors.push(msg.text());
      });
      await page.waitForTimeout(1000);
      if (consoleErrors.length > 0) {
        console.log('Console 错误:', consoleErrors);
      }
      await screenshot('06_guide_not_shown');
    } else {
      console.log('✅ 引导已出现！');
      await screenshot('06_guide_step1');

      // 获取引导文字
      const guideText = await page.evaluate(() => {
        const tooltip = document.getElementById('spotlightTooltip');
        return tooltip ? tooltip.innerText : '未找到';
      });
      console.log('引导内容:', guideText.replace(/\n/g, ' | '));

      // 检查高亮元素
      const highlightedElement = await page.evaluate(() => {
        const overlay = document.getElementById('spotlightOverlay');
        if (!overlay) return null;
        const svg = overlay.querySelector('svg');
        if (!svg) return null;
        const rings = svg.querySelectorAll('.spotlight-ring');
        if (rings.length > 0) {
          const rect = rings[0].getBoundingClientRect();
          // 查找覆盖这个区域的元素
          const elements = document.elementsFromPoint(rect.x + rect.width/2, rect.y + rect.height/2);
          return elements.map(el => el.className || el.id || el.tagName);
        }
        return null;
      });
      console.log('高亮区域覆盖的元素:', highlightedElement);

      // 8. 点"知道了"打开发帖模态框
      const nextBtn = await page.locator('button.spotlight-next, button.spotlight-done').first();
      if (await nextBtn.isVisible()) {
        await nextBtn.click();
        await page.waitForTimeout(500);
        await screenshot('07_post_modal');

        // 9. 填写帖子内容
        await page.fill('#noteInput', '这是引导测试帖子内容！');
        await screenshot('08_post_filled');

        // 10. 提交帖子
        await page.click('button.modal-submit');
        await page.waitForTimeout(1000);
        await screenshot('09_after_post');

        // 11. 检查引导是否更新到 step 2
        const guideAfterPost = await page.evaluate(() => {
          const tooltip = document.getElementById('spotlightTooltip');
          return tooltip ? tooltip.innerText : '未找到';
        });
        console.log('发帖后引导内容:', guideAfterPost.replace(/\n/g, ' | '));
        await screenshot('10_guide_step2');

        console.log('\n✅ 引导流程测试完成！');
      } else {
        console.log('❌ 未找到引导按钮');
      }
    }

  } catch (err) {
    console.error('❌ 测试失败:', err.message);
    await page.screenshot({ path: 'test_error.png' });
    console.log('📸 错误截图: test_error.png');
  } finally {
    await browser.close();
    console.log('\n测试结束');
  }
})();
